File: system-test/dotnet/Dsl.Core/UseCase/External/Tax/UseCases/GetTaxVerification.cs
using Driver.Port.External.Tax.Dtos;
using Dsl.Core.Shared;
using Common;
using Shouldly;

namespace Dsl.Core.External.Tax.UseCases;

public class GetTaxVerification : ResponseVerification<GetTaxResponse>
{
    public GetTaxVerification(GetTaxResponse response, UseCaseContext context)
        : base(response, context)
    {
    }

    public GetTaxVerification Country(string expectedCountryAlias)
    {
        var expectedCountry = Context.GetParamValueOrLiteral(expectedCountryAlias);
        var actualCountry = Response.Country;
        actualCountry.ShouldBe(expectedCountry, $"Expected country to be '{expectedCountry}', but was '{actualCountry}'");
        return this;
    }

    public GetTaxVerification TaxRate(decimal expectedTaxRate)
    {
        var actualTaxRate = Response.TaxRate;
        actualTaxRate.ShouldBe(expectedTaxRate, $"Expected tax rate to be {expectedTaxRate}, but was {actualTaxRate}");
        return this;
    }

    public GetTaxVerification TaxRate(string expectedTaxRate)
    {
        return TaxRate(Converter.ToDecimal(expectedTaxRate)!.Value);
    }

    public GetTaxVerification TaxRateIsPositive()
    {
        var actualTaxRate = Response.TaxRate;
        actualTaxRate.ShouldBeGreaterThan(0m, $"Expected tax rate to be positive, but was {actualTaxRate}");
        return this;
    }
}

File: system-test/java/src/main/java/com/optivem/shop/testkit/dsl/core/usecase/external/tax/usecases/GetTaxVerification.java
package com.optivem.shop.testkit.dsl.core.usecase.external.tax.usecases;

import com.optivem.shop.testkit.common.Converter;
import com.optivem.shop.testkit.dsl.core.shared.ResponseVerification;
import com.optivem.shop.testkit.dsl.core.shared.UseCaseContext;
import com.optivem.shop.testkit.driver.port.external.tax.dtos.GetTaxResponse;

import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThat;

public class GetTaxVerification extends ResponseVerification<GetTaxResponse> {
    public GetTaxVerification(GetTaxResponse response, UseCaseContext context) {
        super(response, context);
    }

    public GetTaxVerification country(String expectedCountryAlias) {
        var expectedCountry = getContext().getParamValueOrLiteral(expectedCountryAlias);
        var actualCountry = getResponse().getCountry();
        assertThat(actualCountry)
                .withFailMessage("Expected country to be '%s', but was '%s'", expectedCountryAlias, actualCountry)
                .isEqualTo(expectedCountry);
        return this;
    }

    public GetTaxVerification taxRate(BigDecimal expectedTaxRate) {
        var actualTaxRate = getResponse().getTaxRate();
        assertThat(actualTaxRate)
                .withFailMessage("Expected tax rate to be %s, but was %s", expectedTaxRate, actualTaxRate)
                .isEqualTo(expectedTaxRate);
        return this;
    }

    public GetTaxVerification taxRate(double expectedTaxRate) {
        return taxRate(Converter.toBigDecimal(expectedTaxRate));
    }

    public GetTaxVerification taxRate(String expectedTaxRate) {
        return taxRate(Converter.toBigDecimal(expectedTaxRate));
    }

    public GetTaxVerification taxRateIsPositive() {
        var actualTaxRate = getResponse().getTaxRate();
        assertThat(actualTaxRate)
                .withFailMessage("Expected tax rate to be positive, but was %s", actualTaxRate)
                .isPositive();
        return this;
    }
}
